README.md
# Roman-ESOLANG-
### **ROMAN** IS AN __ESOTERIC PROGRAMMING LANGUAGE__ THAT USES ROMAN NUMERALS AS A CHARACTER.
REMEMBER THE ROMAN NUMERALS? IF SO, GO TO THE INSTRUCTIONS.

<hr>

IF YOU DON'T KNOW, HERE IS THE TIP:

    • I : 1
    • V : 5
    • X : 10
    • L : 50
    • C : 100
    • D : 500
    • M : 1000

    • IV : 4
    • IX : 9
    • XL : 40
    • XC : 90
    • CD : 400
    • CM : 900

TAKE THE NUMBER IN EXPANDED FORM AND REPLACE IT WITH ROMAN NUMERALS

#### **EXAMPLES**

>     6  => 5 + 1 => V + I => VI
>     14 => 10 + 4 => X + IV => XIV
>     999 => 900 + 90 + 9 => CM + XC + IX => CMXCIX

` NOTE THAT IIII, VVVV, XXXX, LLLL, CCCC, DDDD, MMMM, IIIII, VVVVV, ETC. STILL WORKS IN THE CODE. HOWEVER, IT DOESN'T COUNT AS A ROMAN NUMERALS. `
<hr>

#### **INSTRUCTIONS:**

BEFORE YOU BEGIN THE CODE, PLS PUT THE SOURCE FILE INTO YOUR CODE FOLDER WITH A ROMAN FILE(OR WHATEVER IT IS) OR IN AN EXAMPLE FOLDER.
THE CODE USES SYS' ARGV LIST(INDEX 0) FOR THE TERMINAL. SO THE CODE WILL WORK.
TO START USING THE LANGUAGE, RUN MAIN.PY BECAUSE ITS THE REAL TERMINAL.
PRINTING A STRING OR AN INTEGER IS NOT POSSIBLE IN THIS CODE, INSTEAD, PRINT A CHARACTER.
YOU KNOW THAT EVERY CHARACTER HAS AN ASCII NUMBER,

>     (A -> 65)

IF YOU DON'T KNOW THE NUMBERS OF AN ASCII CHARACTER, LOOK AT THE ASCII TABLE.
TO PRINT A CHARACTER, YOU MUST START WITH A COLON (:)

>     (:CHAR)

NEXT TO IT, PUT A ROMAN NUMBER OF AN ASCII CHARACTER NUMBER. NO SPACES BETWEEN THE COLON AND THE ROMAN NUMERALS.
[IF A IS 65, THEN ITS NUMERAL IS LXV]

>     (:LXV)

CAPITAL LETTERS ARE ONLY ALLOWED FROM THE NUMERALS.
COMMENTS ARE ALLOWED. JUST DONT PUT A COLON AT THE START OF EACH COMMENT

>     (THIS IS A COMMENT)

<hr>

#### EXAMPLE OF THE STRING *"ABC"*:

>     ABC SAMPLE
>     :LXV
>     :LXVI
>     :LXVII
<hr>

#### EXTRA(FOOTER):

`NOTE(IMPORTANT): FILE EXTENSION FOR THIS LANGUAGE IS "rn".`

*HAVE FUN!* :wink:


CHANGELOG.txt
VERSION 1.0.0:
    MAIN:
        - CREATED THIS LANGUAGE

    1.0.1:
        - BUG FIXED

    1.0.2:
        - RENAMED THE FILE "ROMAN1" INTO "ROMAN".
          CAUSING THE CODE TO RAISE AN ERROR ABOUT FILES.
        - FIXED THE PREFIX

    1.0.3:
        - REMOVED THE PRINT(PREFIX) IN TERMINAL.PY

#########################################################

VERSION 1.1.0:
    MAIN:
        - ADDED 'CD' AND 'BACK' COMMANDS
        - FIXED THE BUG ABOUT BYPASSING
          NUMBERS OVER ROMAN NUMERALS.


source/Terminal.py
try:
    from source.Roman import main
except:
    print("RUN ON MAIN.PY")
    quit()
import sys, os

allow_prefix = bool(open(sys.argv[0][:len(sys.argv[0])-12]+"DATA", "r").read())
prefix = sys.argv[0][:len(sys.argv[0])-12].split("/")
prefix = prefix[:len(prefix)-1]


if allow_prefix == True:
    switch = "ON"
else:
    switch = "OFF"

def modcmd(x):
    x = x.split(' ')
    echofold = echopref
    try:
        if x[0] == 'OPEN':
            main("".join(echofold) + x[1])
        elif x[0] == 'CD':
            prefix.append(x[1])
            echofolder = ""
            for folder in prefix:
                echofolder += folder + '/'
            if os.path.isdir(echofolder):
                pass
            else:
                print("FILE TERMINAL ERROR;\nNO DIRECTORY NAMED '{}' EXISTS.".format(echofolder))
                prefix.remove(prefix[len(prefix)-1])
        elif x[0] == 'BACK':
            prefix.remove(prefix[len(prefix)-1])
        else:
            return
    except:
        print("KEYWORD OUT OF RANGE")

if True:
    print("""
    WELCOME TO THE ROMAN1!
    AUTHOR BY OFFICIALAZ3
    VERSION IS 1.1.0
    
    TYPE "ECHO OFF" TO TURN OFF THE ECHO
    TYPE "ECHO ON"  TO TURN ON  THE ECHO
    DEFAULT IS {}
    
    TYPE "CD <FOLDERNAME>" TO OPEN A DIRECTORY
    TYPE "BACK" TO GO BACK
    
    TYPE "EXIT" TO EXIT
    
    TYPE "OPEN <SPACE> <ANY FILE>" TO COMPILE
    
    """.format(switch))
    echopref = ""
    for folder in prefix:
        echopref += folder + '/'
    while True:
        if allow_prefix == True:
            text = input('->{}'.format(echopref))
        else:
            text = input('->')
        print()
            
        if text == "ECHO OFF":
            filename = open(sys.argv[0][:len(sys.argv[0])-12]+"DATA", "w")
            allow_prefix = False
            filename.write("")
            filename.close()
        elif text == "ECHO ON":
            filename = open(sys.argv[0][:len(sys.argv[0])-12]+"DATA", "w")
            allow_prefix = True
            filename.write("HELLO")
            filename.close()
        elif text == "EXIT":
            break
        else:
            modcmd(text)
        print()


source/Roman.py
import sys
    
def hindu(x):
    intsam = 0
    for num in x.split(':'):
        try:
            num = int(num)
            intsam += 1
        except:
            for num n num:
                try:
                    num = int(num)
                    intsam =+ 1
                except:
                    pass
            pass
        if intsam != 0:
            raise SyntaxError
    x = x.replace("IV", ":4").replace("IX", ":9").replace("XL", ":40").replace("XC", ":90").replace("CD", ":400").replace("CM", ":900").replace("I", ":1").replace("V", ":5").replace("X", ":10").replace("L", ":50").replace("C", ":100").replace("D", ":500").replace("M", ":1000").split(':')
    x.remove("")
    sum1 = 0
    for num in x:
        sum1 += int(num)
    return sum1
    
def main(x):
    line = 1
    res = ""
    if not x.endswith('rn'):
        print('FILE ERROR: SHOULD BE IN "RN" FILE EXTENSION')
        return
    try:
        filer = open(x, "r").read().splitlines()
    except:
        print('FILE ERROR: FILE NAMED "{}" NOT FOUND'.format(x))
        return
    for rochar in filer:
        if ':' in rochar:
            try:
                res += chr(hindu(rochar[rochar.index(":")+1:]))
            except:
                print('SYNTAX ERROR: LINE {};\n"{}" -> INVALID ROMAN CODE'.format(line, rochar))
                return
        line += 1
    print(res)

if True:
    try:
        main(sys.argv[1])
    except:
        if __name__ == '__main__':
            print("MUST OPEN MAIN.PY")
